server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.features;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.core.Assertions;
+import org.elasticsearch.plugins.PluginsService;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.NavigableSet;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+/**
+ * Service responsible for registering features this node should publish to other nodes
+ */
+public class FeatureService {
+
+    public static final Version MAX_HISTORICAL_VERSION_EXCLUSIVE = Version.V_8_12_0;
+
+    private static Map<Class<? extends FeatureSpecification>, FeatureSpecification> FEATURE_SPECS;
+
+    private static volatile Set<String> NODE_FEATURES;
+    private static volatile NavigableMap<Version, Set<String>> HISTORICAL_FEATURES;
+
+    private static void readSpecs() {
```

Comment:
nit: the terminology here is confusing, this really should be server features? ie non plugins

Reply:
It's still specs from server, vs those from plugins, right? I think we should have some distinction, otherwise it seems like these are *all* the specs, when they're only some.

Reply:
Not necessarily, in unit tests this will read all the specs as the modules are loaded by the main classpath